### PR TITLE
Codable filesystem implementation - Anton Ilinykh

### PR DESCRIPTION
**Pull request by ailinykh**
Hello guys!
During this challenge I discovered that there are no test for async `insert` cache operation. I mean there are already [assertThatSideEffectsRunSerially](https://github.com/essentialdevelopercom/ios-lead-essentials-feed-store-challenge/blob/master/Tests/FeedStoreSpecs/XCTestCase%2BFeedStoreSpecs.swift#L88:L107) test case but it does not cover the `insert` method without a barrier blocks around it.

You can find the details in [this](https://github.com/essentialdevelopercom/ios-lead-essentials-feed-store-challenge/pull/87/commits/1eaa3a84818a91b81c36a879926299689cd35b67) commit

### CODE REVIEW

File: Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift [hunk 1]
```diff
@@ -106,9 +106,15 @@ extension FeedStoreSpecs where Self: XCTestCase {
 			op3.fulfill()
 		}
 		
+		let op4 = expectation(description: "Operation 4")
+		sut.retrieve(completion: { _ in
+			completedOperationsInOrder.append(op4)
+			op4.fulfill()
+		})
```
Comment (caiozullo) [inline Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift:113]: This extra operation is unnecessary. To pass the challenge, there should be no changes to the provided tests. Please, revert the changes to this file.

File: FeedStoreChallenge/CodableFeedStore.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		queue.async(flags: .barrier) {
+			do {
+				let encoder = JSONEncoder()
+				let cache = Cache(feed: feed.map { CodableFeedImage($0) }, timestamp: timestamp)
+				let data = try encoder.encode(cache)
+				try data.write(to: storeURL)
+			} catch {
+				return completion(error)
+			}
+			completion(nil)
```
Comment (caiozullo) [inline FeedStoreChallenge/CodableFeedStore.swift:72]: the completion should be inside the `do` block to make it easier to follow the code.

File: FeedStoreChallenge/CodableFeedStore.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let storeURL = self.storeURL
+		queue.async(flags: .barrier) {
+			do {
+				let encoder = JSONEncoder()
+				let cache = Cache(feed: feed.map { CodableFeedImage($0) }, timestamp: timestamp)
+				let data = try encoder.encode(cache)
+				try data.write(to: storeURL)
+			} catch {
+				return completion(error)
```
Comment (caiozullo) [inline FeedStoreChallenge/CodableFeedStore.swift:70]: The return is unnecessary if you move the `completion(nil)` inside the `do` block (see my other comment)

**Review by caiozullo — COMMENTED** [review]
Hi @ailinykh, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. Thank you 🙂

File: Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift (hunk 1 above)
Comment (ailinykh) [reply Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift:113]: Hey Caio! Could you read my first comment in this PR please?

Comment (ailinykh) [reply Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift:113]: If you revert [this](https://github.com/essentialdevelopercom/ios-lead-essentials-feed-store-challenge/pull/87/commits/1eaa3a84818a91b81c36a879926299689cd35b67) commit the tests will passed but they should not to

Comment (caiozullo) [reply Tests/FeedStoreSpecs/XCTestCase+FeedStoreSpecs.swift:113]: Hi @ailinykh, the provided tests shouldn't be altered. They are there to validate the desired behavior. But your implementation may need *more tests* (e.g., to simulate concurrency). If that's the case, you should add more tests without altering the existing ones.

**Review by caiozullo — COMMENTED** [review]
Hi @ailinykh,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅ 
- You broke down the implementation into small commits ✅ 
- Bonus points for the Integration Tests ✅ 

Great job!

Feel free to add the tests for the concurrent `insert` method.
